Modify Library to produce a matrix on its own

liornoy opened the pull request:
This commit alters all packages within the library to enable it to generate the matrix autonomously, rather than solely providing utility functions for external matrix creation.

liornoy committed:
Modify Library to produce a matrix on its own

This commit alters all packages within the library to
enable it to generate the matrix autonomously, rather
than solely providing utility functions for external matrix creation.

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+		return nil, fmt.Errorf("failed getting comDetails from `ss` command: %w", err)
+	}
+
+	allComDetails := append(comDetailsFromEndpointSlices, comDetails...)
```
nit: some linters would complain here (assigning and appending with different vars)

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+func getComDetailsFromSS(cs *client.ClientSet, existingComDetails []ComDetails) ([]ComDetails, error) {
+	allComDetails := make([]ComDetails, 0)
```
nit: put this lower before the range

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
general suggestion: run golangci-lint locally (and as part of ci)

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+	res := []ComDetails{}
+	for _, item := range outPuts {
+		str := fmt.Sprintf("%s-%s-%s", item.NodeRole, item.Port, item.Protocol)
+		if _, value := allKeys[str]; !value {
```
nit: can be reduced to `if allKeys[str]` (or use k8s' sets pkg)

liornoy replied:
neat.

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+	var (
+		res   = make([]ComDetails, 0)
+		roles = make(map[string]bool)
+	)
```
nit: `:=` reads better

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+
+func nodeRoles(nodes []corev1.Node, name string) (string, error) {
+	const roleLabel = "node-role.kubernetes.io"
+	var (
```
same

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+	return res
+}
+
+func nodeRoles(nodes []corev1.Node, name string) (string, error) {
```
I think it is cleaner to accept the relevant node instead of filtering inside

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+
+	for role := range roles {
+		for _, p := range epSlice.Ports {
+			ComDetails := ComDetails{
```
nit: just append without assigning to a new var

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+		return "", fmt.Errorf("node %s not found", name)
+	}
+	for label := range node.Labels {
+		if strings.Contains(label, roleLabel) {
+			roles = append(roles, label[len(roleLabel)+1:])
+		}
```
nit: this can be more readable with `strings.CutPrefix` (and a bit more accurate since you expect the role to be the prefix)

liornoy replied:
done

oribon commented on commatrix/commatrix.go:
```diff
@@ -0,0 +1,221 @@
+	}
+
+	sort.Strings(roles)
+	cleanedRoles := filterValidRoles(roles)
```
nit: instead of a function this can be a part of the loop (with a map containing the valid values)

liornoy replied:
done

oribon commented on internal/client/client.go:
```diff
@@ -19,24 +19,22 @@ type ClientSet struct {
 
-	if kubeconfig == "" {
-		kubeconfig = os.Getenv("KUBECONFIG")
+	if kubeconfigPath == "" {
+		if kubeconfigPath, ok = os.LookupEnv("KUBECONFIG"); !ok {
+			return nil, errors.New("no kubeconfig found")
+		}
 	}
```
nit: I'd expect the caller to make these assumptions and not this inner func

liornoy replied:
done

oribon commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,165 @@
+// and returns the DebugPod object. Use the Clean() method to delete it.
+func New(node string) (*DebugPod, error) {
+	errCh := make(chan error)
+	go func() {
+		errCh <- start(node)
+	}()
+	for {
+		select {
+		case err := <-errCh:
+			return nil, err
+		default:
+			return waitForPod()
+		}
+	}
```
this seems a bit odd, what is the rationale of not making this "synchronous" (start -> wait)?

liornoy replied:
yes, it's not trivial and took me a couple of tries to figure it out.
It's just that the start func goes to sleep together with the "oc debug... sleep inf.."
So it must be inside a go routine

liornoy replied:
After a private discussion, we agreed to explore a different approach, possibly leveraging an existing privileged pod instead, as this code introduces a race condition.

oribon commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,165 @@
+	return out, nil
+}
+
+func (dp *DebugPod) ExecWithRetry(cmd string) ([]byte, error) {
```
nit: I think you should accept the interval/duration here and not hardcode it

liornoy replied:
done

oribon commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,165 @@
+		return fmt.Errorf("failed deleting debug pod %s/%s: %v\n%s", dp.Namespace, dp.Name, err, string(output))
+	}
+
+	output, err = exec.Command("oc", "delete", "ns", dp.Namespace).CombinedOutput()
```
why delete the entire namespace? (does each pod gets its own ns?)

liornoy replied:
yes

liornoy replied:
however, maybe it's better to set the namespace myself
because I found out that if the "oc" tool set the working project to something,
it will create them in the current project. and that will fail everything.

oribon commented on internal/debug/debug.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// get returns the debug pod that is present in the cluster
+func get() (*DebugPod, error) {
```
I think this function could be simplified by using a client and not oc

liornoy replied:
Yep, I agree

oribon commented on internal/nftables/nftables.go:
```diff
@@ -0,0 +1,27 @@
+package nftables
+
+type NftablesData struct {
```
nit: can be renamed to Data

liornoy replied:
done

oribon commented on internal/nftables/nftables.go:
```diff
@@ -0,0 +1,27 @@
+	AllowedUDPPorts []string
+}
+
+const NftablesTemplate = `
```
nit: can be renamed to Template

liornoy replied:
done

liornoy committed:
Add golangci linter check to CI/CD

This commit adds a new step in the CI/CD process,
incorporating the golangci linter to analyze the repository.

As part of this update, the Makefile is eliminated
and we transition to using tasks.py.

Signed-off-by: Lior Noy <lnoy@redhat.com>

liornoy committed:
Move logic for custom endpointslices to new module

This commit relocates the logic responsible for generating custom
endpointslices to a dedicated module.
The main `commatrix` logic now invokes `customEndpointSlices` to
generate services that aren't natively represented by endpoint slices.

The module utilizes the "ss" library to extract listening ports on
the nodes' machines. The `FilterPorts` function is employed to retrieve
only known port numbers or service names. Subsequently, endpoint slices
are created on the cluster with the temporary prefix "commatrix-test-"
to indicate their custom nature.

The list of known ports resides in "internal/ss/ports.go" and is sourced
from the OpenShift enhancements repository:

https://github.com/openshift/enhancements/blob/master/dev-guide/host-port-registry.md

Signed-off-by: Lior Noy <lnoy@redhat.com>

liornoy committed:
Remove pointer module and use k8s/utils/ptr

Signed-off-by: Lior Noy <lnoy@redhat.com>

liornoy committed:
Refactor the query module to filter

Modified query module to filter module.
This to improve readability, and removed
the "WithLabel" because the requirements changed
and we will not create custom endpointslices.

liornoy committed:
modify query to filter

liornoy committed:
Rename internal folder to pkg

We rename the folder to "pkg" instead of "internal"
so that we will be able to use the helped functions
of different modules outside of commatrix.

liornoy committed:
Add Remove Duplications funciton

liornoy committed:
Add Diff funciton

liornoy committed:
Change ComDeatils structure

liornoy committed:
Adding new fields to ComDetails

This commit adds new fields to comDetails, and
deletes "matrixcustomizer" module to only
allow using a file path.

liornoy committed:
Update static-sutom-entries.json

liornoy committed:
Add static entries

liornoy committed:
fix kubelet entry

liornoy committed:
Improve log messages

Signed-off-by: Lior Noy <lnoy@redhat.com>

liornoy committed:
Introduce the env enum for static entries

Added the env enum to allow selection between different environments
for the cluster (BareMetal / AWS).

This splits the previous staticEntries variable into generic, BM, and
AWS, to allow flexibility in choosing the target infrastructure.

Signed-off-by: Lior Noy <lnoy@redhat.com>

liornoy committed:
Add csv header line

liornoy commented:
closing this in favor of openshift-kni/commatrix repo
